server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/FilterAllocationDecider.java

```diff
@@ -102,6 +102,17 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
         return shouldFilter(shardRouting, node.node(), allocation);
     }
 
+    @Override
+    public Decision canForceDuringVacate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        if (replacementFromSourceToTarget(allocation.metadata(), shardRouting.currentNodeId(), node.nodeId())) {
```

Comment:
Also callers responsibility to only call this when a replace is ongoing between source and target.